README.md
# swift-coverage

Display coverage reports next to your code for Swift Package Manager projects.

![](./assets/docs/example.png)

To update and show your coverage statistics, run your tests with code coverage enabled.

```bash
$ swift test --enable-code-coverage
```

## Configuration

| Config                                        | Type                    | Default                                       | Description                                                             |
| --------------------------------------------- | ----------------------- | --------------------------------------------- | ----------------------------------------------------------------------- |
| swift-coverage.coverageFilePattern            | `string`; glob pattern  | `**/.build/*/debug/{coverage,codecov}/*.json` | The pattern where to find the llvm-codecov JSON-reports.                |
| swift-coverage.highlightsCoveredText          | `boolean`               | `false`                                       | Wether code with coverage should have a higlighted background.          |
| swift-coverage.highlightsNotCoveredText       | `boolean`               | `true`                                        | Wether code with no coverage should have a higlighted background.       |
| swift-coverage.coveredTextBackground.dark     | `string`, CSS-color     | `rgba(48, 209, 88, 0.25)`                     | The brackground color used to highlight code with coverage.             |
| swift-coverage.coveredTextBackground.light    | `string`, CSS-color     | `rgba(52, 199, 89, 0.25)`                     | The brackground color used to highlight code with coverage.             |
| swift-coverage.notCoveredTextBackground.dark  | `string`, CSS-color     | `rgba(255, 69, 58, 0.25)`                     | The brackground color used to highlight code with no coverage.          |
| swift-coverage.notCoveredTextBackground.light | `string`, CSS-color     | `rgba(255, 59, 48, 0.25)`                     | The brackground color used to highlight code with no coverage.          |
| swift-coverage.highlightsCoveredGutter        | `boolean`               | `true`                                        | Wether lines with coverage should have a highlight-gutter.              |
| swift-coverage.highlightsNotCoveredGutter     | `boolean`               | `true`                                        | Wether lines with no coverage should have a highlight-gutter.           |
| swift-coverage.highlightsMixedGutter          | `boolean`               | `true`                                        | Wether lines with mixed coverage should have a highlight-gutter.        |
| swift-coverage.coveredGutterIcon.dark         | `string`, absolute path | green indicator (builtin)                     | The path to the gutter-icon used to highlight code with coverage.       |
| swift-coverage.coveredGutterIcon.light        | `string`, absolute path | green indicator (builtin)                     | The path to the gutter-icon used to highlight code with coverage.       |
| swift-coverage.notCoveredGutterIcon.dark      | `string`, absolute path | red indicator (builtin)                       | The path to the gutter-icon used to highlight code without coverage.    |
| swift-coverage.notCoveredGutterIcon.light     | `string`, absolute path | red indicator (builtin)                       | The path to the gutter-icon used to highlight code without coverage.    |
| swift-coverage.mixedGutterIcon.dark           | `string`, absolute path | orange indicator (builtin)                    | The path to the gutter-icon used to highlight code with mixed coverage. |
| swift-coverage.mixedGutterIcon.light          | `string`, absolute path | orange indicator (builtin)                    | The path to the gutter-icon used to highlight code with mixed coverage. |

## Contributors

- Valentin Knabel, [@vknabel](https://github.com/vknabel), [@vknabel](https://twitter.com/vknabel) on Twitter

## License

vscode-swift-coverage is available under the [MIT](./LICENSE) license.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.1] - 2020-12-11

### Fixed

- Included missing example image in VS Code Extension bundle
- Also accepts default `coverage`
- The config `swift-coverage.coverageFilePattern` was ignored

## [1.0.0] - 2020-08-29

- Initial release


src/llvm-cov.ts
export interface LLVMCov {
  type: string;
  version: string;
  data: LLVMCovData[];
}

export interface LLVMCovData {
  files: LLVMCovFile[];
  functions: LLVMCovFunction[];
  totals: LLVMCovTotals;
}

export interface LLVMCovTotals {
  functions: LLVMCovSummary;
  instantiations: LLVMCovSummary;
  lines: LLVMCovSummary;
  regions: LLVMCovSummary & { notcovered: number };
}

export interface LLVMCovSummary {
  count: number;
  covered: number;
  /**
   * Between 0 and 100; floating point
   */
  percent: number;
}

export interface LLVMCovFile {
  expansions: unknown[];
  filename: string;
  /**
   * [Line, Column, Executions, hasCoverage, ?]
   */
  segments: Array<[number, number, number, boolean, boolean]>;
  summary: LLVMCovTotals;
}

export interface LLVMCovFunction {
  count: number;
  filenames: string[];
  /**
   * Mangled name
   */
  name: string;
  regions: Array<
    [number, number, number, number, number, number, number, number]
  >;
}


src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from "vscode";
import { currentConfig } from "./config";
import { CoverageWatcher } from "./coverage-watcher";

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
  let coverageWatcher = new CoverageWatcher(currentConfig(context));
  context.subscriptions.push(coverageWatcher);

  vscode.workspace.onDidChangeConfiguration(() => {
    coverageWatcher.dispose();

    coverageWatcher = new CoverageWatcher(currentConfig(context));
    context.subscriptions.push(coverageWatcher);
  });
}

// this method is called when your extension is deactivated
export function deactivate() {}


src/coverage-watcher.ts
import * as fs from "fs";
import { Disposable } from "vscode";
import { TotalCoverageReporter } from "./total-coverage-reporter";
import { EditorCoverageReporter } from "./editor-coverage-reporter";
import { Config } from "./config";
import { LLVMCov } from "./llvm-cov";
import { promisify } from "util";
import * as vscode from "vscode";

export class CoverageWatcher implements Disposable {
  private readonly coverageReporter: TotalCoverageReporter;
  private readonly editorCoverageReporter: EditorCoverageReporter;

  private readonly subscriptions = Array<Disposable>();

  constructor(config: Config) {
    this.coverageReporter = new TotalCoverageReporter();
    this.editorCoverageReporter = new EditorCoverageReporter(config);

    this.subscriptions.push(
      this.coverageReporter,
      this.editorCoverageReporter,
      vscode.workspace
        .createFileSystemWatcher(config.coverageFilePattern)
        .onDidChange((uri) => this.applyCoverage(uri)),
      vscode.workspace
        .createFileSystemWatcher(config.coverageFilePattern)
        .onDidCreate((uri) => this.applyCoverage(uri))
    );

    vscode.workspace
      .findFiles(config.coverageFilePattern)
      .then((uris) => uris.forEach((uri) => this.applyCoverage(uri)));
  }

  private async applyCoverage(file: vscode.Uri) {
    let llvmCov: LLVMCov;
    if (file.scheme == "file") {
      const document = await vscode.workspace.openTextDocument(file);
      llvmCov = JSON.parse(document.getText());
    } else {
      llvmCov = JSON.parse(await promisify(fs.readFile)(file.fsPath, "utf8"));
    }

    this.coverageReporter.display(llvmCov);
    this.editorCoverageReporter.display(llvmCov);
  }

  dispose() {
    this.subscriptions.forEach((subscription) => subscription.dispose());
  }
}


src/total-coverage-reporter.ts
import { Disposable } from "vscode";
import * as vscode from "vscode";
import { LLVMCov, LLVMCovSummary, LLVMCovTotals } from "./llvm-cov";

export class TotalCoverageReporter implements Disposable {
  private readonly coverageStatusBarItem = vscode.window.createStatusBarItem(
    vscode.StatusBarAlignment.Right,
    100
  );
  private isShown = false;

  display(llvmCov: LLVMCov) {
    if (llvmCov.data.length == 0) {
      this.coverageStatusBarItem.hide();
      return;
    }
    const totals = llvmCov.data[0].totals;
    this.coverageStatusBarItem.text = this.summarizeForBarItem(totals.lines);
    this.coverageStatusBarItem.tooltip = [
      this.summarizeTotalsForTooltip("Lines", "lines", totals),
      this.summarizeTotalsForTooltip("Functions", "functions", totals),
      this.summarizeTotalsForTooltip(
        "Instantiations",
        "instantiations",
        totals
      ),
      this.summarizeTotalsForTooltip("Regions", "regions", totals),
    ].join("\n");

    if (!this.isShown) {
      this.coverageStatusBarItem.show();
    }
  }

  private summarizeTotalsForTooltip(
    label: string,
    total: keyof LLVMCovTotals,
    totals: LLVMCovTotals
  ): string {
    return `${label}: ${this.summarizeForTooltip(totals[total])}`;
  }

  private summarizeForTooltip(summary: LLVMCovSummary): string {
    return `${summary.percent.toFixed()}% (${summary.covered}/${
      summary.count
    })`;
  }
  private summarizeForBarItem(summary: LLVMCovSummary): string {
    return `${summary.percent.toFixed()}%`;
  }

  dispose() {
    this.coverageStatusBarItem.dispose();
  }
}


src/document-range-decorator.ts
import * as vscode from "vscode";

export interface CoverageDecoration {
  kind: "covered" | "notcovered" | "mixed";
  options: vscode.DecorationOptions;
}

export class DocumentRangeDecorator {
  private readonly allDecorations = new Array<CoverageDecoration>();
  private readonly lineDecorations = new Array<CoverageDecoration[]>();

  addDecoration(decoration: CoverageDecoration) {
    this.allDecorations.push(decoration);

    for (
      let lineNumber = decoration.options.range.start.line;
      lineNumber <= decoration.options.range.end.line;
      lineNumber++
    ) {
      const existingDecorations = this.lineDecorations[lineNumber] || Array();
      this.lineDecorations[lineNumber] = [...existingDecorations, decoration];
    }
  }

  textDecorationsForCoverageKind(
    kind: CoverageDecoration["kind"]
  ): vscode.DecorationOptions[] {
    return this.allDecorations
      .filter((decoration) => decoration.kind === kind)
      .map(({ options }) => options);
  }

  gutterDecorationsForCoverageKind(
    kind: CoverageDecoration["kind"]
  ): vscode.DecorationOptions[] {
    return this.gutterCoverageDecorations()
      .filter((decoration) => decoration.kind === kind)
      .map(({ options }) => options);
  }

  private gutterCoverageDecorations(): CoverageDecoration[] {
    return this.lineDecorations
      .map(
        (line, index): CoverageDecoration =>
          line && {
            kind: this.gutterKindForCoverageDecorations(line),
            options: { range: new vscode.Range(index, 0, index, 1) },
          }
      )
      .filter((gutter) => !!gutter);
  }

  private gutterKindForCoverageDecorations(
    line: CoverageDecoration[]
  ): CoverageDecoration["kind"] {
    const kinds = line.map(({ kind }) => kind);
    if (
      kinds.includes("mixed") ||
      (kinds.includes("covered") && kinds.includes("notcovered"))
    ) {
      return "mixed";
    } else {
      return kinds.includes("covered") ? "covered" : "notcovered";
    }
  }
}


src/editor-coverage-reporter.ts
import { Disposable } from "vscode";
import * as vscode from "vscode";
import { LLVMCov, LLVMCovFile } from "./llvm-cov";
import { DocumentRangeDecorator } from "./document-range-decorator";
import { Config } from "./config";

export class EditorCoverageReporter implements Disposable {
  private readonly editorChanges = vscode.window.onDidChangeVisibleTextEditors(
    (editors) => this.visibleEditorsDidChange(editors)
  );
  private readonly xeditorChanges = vscode.window.onDidChangeActiveTextEditor(
    (editor) => editor && this.visibleEditorsDidChange([editor])
  );
  private readonly gutterCoveredDecoration: vscode.TextEditorDecorationType;
  private readonly gutterNotCoveredDecoration: vscode.TextEditorDecorationType;
  private readonly gutterMixedDecoration: vscode.TextEditorDecorationType;

  private readonly textCoveredDecoration: vscode.TextEditorDecorationType;
  private readonly textNotCoveredDecoration: vscode.TextEditorDecorationType;
  private readonly cache = new Map<string, LLVMCovFile>();

  constructor(private readonly config: Config) {
    this.textCoveredDecoration = vscode.window.createTextEditorDecorationType({
      dark: {
        backgroundColor: config.coveredTextBackground.dark,
      },
      light: {
        backgroundColor: config.coveredTextBackground.light,
      },
    });
    this.textNotCoveredDecoration = vscode.window.createTextEditorDecorationType(
      {
        dark: {
          backgroundColor: config.notCoveredTextBackground.dark,
        },
        light: {
          backgroundColor: config.notCoveredTextBackground.light,
        },
      }
    );

    this.gutterCoveredDecoration = vscode.window.createTextEditorDecorationType(
      {
        dark: {
          gutterIconPath: config.coveredGutterIcon.dark,
        },
        light: {
          gutterIconPath: config.coveredGutterIcon.light,
        },
      }
    );
    this.gutterNotCoveredDecoration = vscode.window.createTextEditorDecorationType(
      {
        dark: {
          gutterIconPath: config.notCoveredGutterIcon.dark,
        },
        light: {
          gutterIconPath: config.notCoveredGutterIcon.light,
        },
      }
    );
    this.gutterMixedDecoration = vscode.window.createTextEditorDecorationType({
      dark: {
        gutterIconPath: config.mixedGutterIcon.dark,
      },
      light: {
        gutterIconPath: config.mixedGutterIcon.light,
      },
    });
  }

  display(llvmCov: LLVMCov) {
    if (llvmCov.data.length == 0) {
      return;
    }
    this.cache.clear();
    llvmCov.data.forEach((data) => {
      data.files.forEach((file) => {
        this.cache.set(file.filename, file);
      });
    });
    // TODO: Render
    this.visibleEditorsDidChange(vscode.window.visibleTextEditors);
  }

  visibleEditorsDidChange(editors: vscode.TextEditor[]): Disposable {
    // TODO: remove from editors?
    for (const editor of editors) {
      const fileCov = this.cache.get(editor.document.uri.fsPath);
      if (!fileCov) {
        return { dispose: () => void 0 };
      }

      const documentRangeDecorator = new DocumentRangeDecorator();
      for (let i = 0; i < fileCov.segments.length; i++) {
        const [
          line,
          column,
          executions,
          hasCoverage,
          _notsure,
        ] = fileCov.segments[i];
        if (!hasCoverage) {
          continue;
        }

        const [endLine, endColumn] =
          fileCov.segments[i + 1] ||
          editor.document.lineAt(editor.document.lineCount - 1);

        const startPosition = new vscode.Position(line - 1, column - 1);
        const endPosition = new vscode.Position(endLine - 1, endColumn);

        const decoration: vscode.DecorationOptions = {
          range: new vscode.Range(startPosition, endPosition),
          renderOptions: {
            // before: {
            //   contentText: `${executions}`,
            // },
            // after: {
            //   contentText: `${executions}`,
            // },
          },
          hoverMessage: `*Executed ${
            executions == 1 ? "once" : `${executions} times`
          } in tests*`,
          // hoverMessage
          // renderOptions: {
          //   before: {
          //     contentText: "Hello",
          //   },
          // },
        };

        documentRangeDecorator.addDecoration({
          kind: executions > 0 ? "covered" : "notcovered",
          options: decoration,
        });

        if (this.config.highlightsCoveredGutter) {
          editor.setDecorations(
            this.gutterCoveredDecoration,
            documentRangeDecorator.gutterDecorationsForCoverageKind("covered")
          );
        }
        if (this.config.highlightsNotCoveredGutter) {
          editor.setDecorations(
            this.gutterNotCoveredDecoration,
            documentRangeDecorator.gutterDecorationsForCoverageKind(
              "notcovered"
            )
          );
        }
        if (this.config.highlightsMixedGutter) {
          editor.setDecorations(
            this.gutterMixedDecoration,
            documentRangeDecorator.gutterDecorationsForCoverageKind("mixed")
          );
        }

        if (this.config.highlightsCoveredText) {
          editor.setDecorations(
            this.textCoveredDecoration,
            documentRangeDecorator.textDecorationsForCoverageKind("covered")
          );
        }
        if (this.config.highlightsNotCoveredText) {
          editor.setDecorations(
            this.textNotCoveredDecoration,
            documentRangeDecorator.textDecorationsForCoverageKind("notcovered")
          );
        }
      }
    }
    return { dispose: () => void 0 };
  }

  dispose() {
    this.editorChanges.dispose();
    this.gutterCoveredDecoration.dispose();
    this.gutterNotCoveredDecoration.dispose();
    this.gutterMixedDecoration.dispose();
    this.textCoveredDecoration.dispose();
    this.textNotCoveredDecoration.dispose();
  }
}


src/config.ts
import * as vscode from "vscode";

export interface DarkAndLight<T> {
  dark: T;
  light: T;
}

export interface Config {
  coverageFilePattern: string;

  highlightsCoveredText: boolean;
  highlightsNotCoveredText: boolean;

  coveredTextBackground: DarkAndLight<string>;
  notCoveredTextBackground: DarkAndLight<string>;

  highlightsCoveredGutter: boolean;
  highlightsNotCoveredGutter: boolean;
  highlightsMixedGutter: boolean;

  coveredGutterIcon: DarkAndLight<string>;
  notCoveredGutterIcon: DarkAndLight<string>;
  mixedGutterIcon: DarkAndLight<string>;
}

export function currentConfig(context: vscode.ExtensionContext): Config {
  const conf = vscode.workspace.getConfiguration("swift-coverage");
  return {
    coverageFilePattern: conf.get<string>(
      "coverageFilePattern",
      "**/.build/*/debug/{coverage,codecov}/*.json"
    )!,

    highlightsCoveredText: conf.get("highlightsCoveredText", false),
    highlightsNotCoveredText: conf.get("highlightsNotCoveredText", true),

    coveredTextBackground: {
      dark: conf.get("coveredTextBackground.dark", "rgba(48, 209, 88, 0.25)"),
      light: conf.get("coveredTextBackground.light", "rgba(52, 199, 89, 0.25)"),
    },
    notCoveredTextBackground: {
      dark: conf.get(
        "notCoveredTextBackground.dark",
        "rgba(255, 69, 58, 0.25)"
      ),
      light: conf.get(
        "notCoveredTextBackground.light",
        "rgba(255, 59, 48, 0.25)"
      ),
    },

    highlightsCoveredGutter: conf.get("highlightsCoveredGutter", true),
    highlightsNotCoveredGutter: conf.get("highlightsNotCoveredGutter", true),
    highlightsMixedGutter: conf.get("highlightsMixedGutter", true),

    coveredGutterIcon: {
      dark:
        conf.get("coveredGutterIcon.dark") ||
        context.asAbsolutePath("./assets/gutter-icon--covered-dark.svg"),
      light:
        conf.get("coveredGutterIcon.light") ||
        context.asAbsolutePath("./assets/gutter-icon--covered-light.svg"),
    },
    notCoveredGutterIcon: {
      dark:
        conf.get("notCoveredGutterIcon.dark") ||
        context.asAbsolutePath("./assets/gutter-icon--notcovered-dark.svg"),
      light:
        conf.get("notCoveredGutterIcon.light") ||
        context.asAbsolutePath("./assets/gutter-icon--notcovered-light.svg"),
    },
    mixedGutterIcon: {
      dark:
        conf.get("mixedGutterIcon.dark") ||
        context.asAbsolutePath("./assets/gutter-icon--mixed-dark.svg"),
      light:
        conf.get("mixedGutterIcon.light") ||
        context.asAbsolutePath("./assets/gutter-icon--mixed-light.svg"),
    },
  };
}
